Need `#include <algorithm>` to build `gpt4all-backend/llamamodel.cpp`

### Bug Report

When building gpt4all-chat on ArchLinux from the [AUR](https://aur.archlinux.org/packages/gpt4all-chat), the compiler fails with an error message about `transform`, `find`, `find_if` not being a member of `std`:

```
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:892:22: error: ‘transform’ is not a member of ‘std’
  892 |                 std::transform(embd, embd_end, embd, [mean](double f){ return f - mean; });
      |                      ^~~~~~~~~
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:901:22: error: ‘transform’ is not a member of ‘std’
  901 |                 std::transform(embd, embd_end, embd, product(1.0 / std::sqrt(variance + 1e-5)));
      |                      ^~~~~~~~~
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:906:18: error: ‘transform’ is not a member of ‘std’
  906 |             std::transform(embd, embd_end, out, out, [scale](double e, double o){ return o + scale * e; });
      |                  ^~~~~~~~~
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp: In member function ‘void LLamaModel::embedInternal(const std::vector<std::__cxx11::basic_string<char> >&, float*, std::string, int, size_t*, bool, bool, bool (*)(unsigned int*, unsigned int, const char*), const EmbModelSpec*)’:
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:934:14: error: ‘transform’ is not a member of ‘std’
  934 |         std::transform(embd, embd_end, embd, product(1.0 / total));
      |              ^~~~~~~~~
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:938:14: error: ‘transform’ is not a member of ‘std’
  938 |         std::transform(embd, embd_end, embeddings, product(scale));
      |              ^~~~~~~~~
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp: In function ‘bool is_arch_supported(const char*)’:
/home/user/.cache/paru/clone/gpt4all-chat/src/gpt4all-2.7.5/gpt4all-backend/llamamodel.cpp:980:21: error: no matching function for call to ‘find(std::vector<const char*>::const_iterator, std::vector<const char*>::const_iterator, std::string)’
  980 |     return std::find(KNOWN_ARCHES.begin(), KNOWN_ARCHES.end(), std::string(arch)) < KNOWN_ARCHES.end();
      |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### Steps to Reproduce

(ArchLinux steps only, it may be due to the compiler version that comes with `base-devel`)

1. Clone the `gpt4all-chat` packaging repository from AUR: `git clone https://aur.archlinux.org/gpt4all-chat.git`
2. Change into the cloned directory: `cd gpt4all-chat`
3. Run `makepkg -si` to build and install the package

### Expected Behavior

To resolve this issue, I added a single line of code (`#include <algorithm>`) at the top of the `llamamodel.cpp` file. This patch allows compilation to succeed.

I reported this potential Arch Linux-specific bug to the AUR package maintainer and found that similar issues have been discussed on Stack Overflow (see [1](https://stackoverflow.com/questions/10227928/build-error-with-c-find-if-is-not-a-member-of-std) and [2](https://stackoverflow.com/questions/57136998/transform-is-not-a-member-of-std)). These threads suggest that including `<algorithm>` is a common solution for these types of errors.

### Your Environment

- GPT4All version (if applicable): 2.7.5
- Operating System: ArchLinux
- Chat model used (if applicable): N/A

--- a/gpt4all-backend/CMakeLists.txt
+++ b/gpt4all-backend/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.16)
+cmake_minimum_required(VERSION 3.21)  # for PROJECT_IS_TOP_LEVEL
 set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
@@ -141,7 +141,7 @@ foreach(BUILD_VARIANT IN LISTS BUILD_VARIANTS)
         prepare_target(gptj llama-mainline)
     endif()
 
-    if (BUILD_VARIANT STREQUAL cuda)
+    if (NOT PROJECT_IS_TOP_LEVEL AND BUILD_VARIANT STREQUAL cuda)
         set(CUDAToolkit_BIN_DIR ${CUDAToolkit_BIN_DIR} PARENT_SCOPE)
     endif()
 endforeach()

--- a/gpt4all-backend/gptj.cpp
+++ b/gpt4all-backend/gptj.cpp
@@ -5,6 +5,8 @@
 #include "llmodel_shared.h"
 #include "utils.h"
 
+#include <ggml.h>
+
 #include <algorithm>
 #include <cassert>
 #include <cinttypes>
@@ -21,9 +23,6 @@
 #include <string>
 #include <thread>
 #include <vector>
-
-#include <ggml.h>
-
 
 namespace {
 const char *modelType_ = "GPT-J";

--- a/gpt4all-backend/gptj.cpp
+++ b/gpt4all-backend/gptj.cpp
@@ -1,31 +1,27 @@
 #define GPTJ_H_I_KNOW_WHAT_I_AM_DOING_WHEN_INCLUDING_THIS_FILE
 #include "gptj_impl.h"
 
-#include "utils.h"
+#include "llmodel.h"
 #include "llmodel_shared.h"
+#include "utils.h"
 
+#include <algorithm>
 #include <cassert>
 #include <cinttypes>
 #include <cmath>
 #include <cstdio>
 #include <cstring>
-#include <map>
-#include <string>
-#include <vector>
+#include <ctime>
 #include <iostream>
-#if defined(_WIN32) && defined(_MSC_VER)
-    #define WIN32_LEAN_AND_MEAN
-    #ifndef NOMINMAX
-        #define NOMINMAX
-    #endif
-    #include <windows.h>
-    #include <io.h>
-    #include <stdio.h>
-#else
-    #include <unistd.h>
-#endif
+#include <map>
+#include <memory>
+#include <random>
 #include <sstream>
-#include <unordered_set>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <vector>
+
 #include <ggml.h>
 
 

--- a/gpt4all-backend/gptj_impl.h
+++ b/gpt4all-backend/gptj_impl.h
@@ -4,10 +4,11 @@
 #ifndef GPTJ_H
 #define GPTJ_H
 
-#include <string>
-#include <functional>
-#include <vector>
 #include "llmodel.h"
+
+#include <functional>
+#include <string>
+#include <vector>
 
 struct GPTJPrivate;
 class GPTJ : public LLModel {

--- a/gpt4all-backend/llamamodel.cpp
+++ b/gpt4all-backend/llamamodel.cpp
@@ -3,6 +3,9 @@
 
 #include "llmodel.h"
 
+#include <ggml.h>
+#include <llama.h>
+
 #include <algorithm>
 #include <cassert>
 #include <cmath>
@@ -25,8 +28,6 @@
 #include <thread>
 #include <vector>
 
-#include <llama.h>
-#include <ggml.h>
 #ifdef GGML_USE_KOMPUTE
 #   include <ggml-kompute.h>
 #elif GGML_USE_VULKAN

--- a/gpt4all-backend/llamamodel.cpp
+++ b/gpt4all-backend/llamamodel.cpp
@@ -1,22 +1,28 @@
 #define LLAMAMODEL_H_I_KNOW_WHAT_I_AM_DOING_WHEN_INCLUDING_THIS_FILE
 #include "llamamodel_impl.h"
 
+#include "llmodel.h"
+
+#include <algorithm>
 #include <cassert>
 #include <cmath>
+#include <cstdint>
 #include <cstdio>
+#include <cstdlib>
 #include <cstring>
 #include <fstream>
+#include <functional>
 #include <initializer_list>
 #include <iomanip>
 #include <iostream>
-#include <map>
+#include <iterator>
+#include <memory>
 #include <numeric>
-#include <random>
+#include <optional>
 #include <sstream>
 #include <stdexcept>
 #include <string>
 #include <thread>
-#include <unordered_set>
 #include <vector>
 
 #include <llama.h>
@@ -30,6 +36,7 @@
 #endif
 
 using namespace std::string_literals;
+
 
 // Maximum supported GGUF version
 static constexpr int GGUF_VER_MAX = 3;

--- a/gpt4all-backend/llamamodel_impl.h
+++ b/gpt4all-backend/llamamodel_impl.h
@@ -4,11 +4,12 @@
 #ifndef LLAMAMODEL_H
 #define LLAMAMODEL_H
 
+#include "llmodel.h"
+
 #include <functional>
 #include <memory>
 #include <string>
 #include <vector>
-#include "llmodel.h"
 
 struct LLamaPrivate;
 struct EmbModelSpec;

--- a/gpt4all-backend/llmodel.cpp
+++ b/gpt4all-backend/llmodel.cpp
@@ -1,12 +1,12 @@
-#include "llmodel.h"
 #include "dlhandle.h"
-#include "sysinfo.h"
+#include "llmodel.h"
 
 #include <cassert>
 #include <cstdlib>
 #include <filesystem>
 #include <fstream>
 #include <iostream>
+#include <iterator>
 #include <memory>
 #include <optional>
 #include <regex>
@@ -25,6 +25,10 @@
 
 #ifdef _MSC_VER
 #   include <intrin.h>
+#endif
+
+#if defined(__APPLE__) && defined(__aarch64__)
+#   include "sysinfo.h" // for getSystemTotalRAMInBytes
 #endif
 
 #ifndef __APPLE__

--- a/gpt4all-backend/llmodel.h
+++ b/gpt4all-backend/llmodel.h
@@ -2,14 +2,15 @@
 #define LLMODEL_H
 
 #include <algorithm>
+#include <cstddef>
 #include <cstdint>
-#include <fstream>
 #include <functional>
-#include <limits>
 #include <optional>
+#include <stdexcept>
 #include <string>
 #include <string_view>
 #include <unordered_map>
+#include <utility>
 #include <vector>
 
 using namespace std::string_literals;

--- a/gpt4all-backend/llmodel_c.cpp
+++ b/gpt4all-backend/llmodel_c.cpp
@@ -1,12 +1,17 @@
-#include "llmodel_c.h"
 #include "llmodel.h"
+#include "llmodel_c.h"
 
-#include <cerrno>
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
 #include <cstring>
+#include <exception>
+#include <functional>
 #include <iostream>
 #include <memory>
 #include <optional>
-#include <utility>
+#include <string>
+#include <vector>
 
 struct LLModelWrapper {
     LLModel *llModel = nullptr;

--- a/gpt4all-backend/llmodel_c.cpp
+++ b/gpt4all-backend/llmodel_c.cpp
@@ -1,5 +1,6 @@
-#include "llmodel.h"
 #include "llmodel_c.h"
+
+#include "llmodel.h"
 
 #include <algorithm>
 #include <cstdio>

--- a/gpt4all-backend/llmodel_c.h
+++ b/gpt4all-backend/llmodel_c.h
@@ -1,9 +1,9 @@
 #ifndef LLMODEL_C_H
 #define LLMODEL_C_H
 
-#include <stdint.h>
-#include <stddef.h>
 #include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 #ifdef __GNUC__
 #define DEPRECATED __attribute__ ((deprecated))

--- a/gpt4all-backend/llmodel_shared.cpp
+++ b/gpt4all-backend/llmodel_shared.cpp
@@ -1,10 +1,17 @@
 #include "llmodel.h"
 
+#include <algorithm>
 #include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <functional>
 #include <iostream>
+#include <optional>
 #include <regex>
+#include <stdexcept>
 #include <string>
 #include <unordered_set>
+#include <vector>
 
 // TODO(cebtenzzre): replace this with llama_kv_cache_seq_shift for llamamodel (GPT-J needs this as-is)
 void LLModel::recalculateContext(PromptContext &promptCtx, std::function<bool(bool)> recalculate) {

--- a/gpt4all-backend/llmodel_shared.h
+++ b/gpt4all-backend/llmodel_shared.h
@@ -1,8 +1,10 @@
 #pragma once
-#include <cstdint>
-#include <cstddef>
-#include <vector>
+
 #include <ggml.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <vector>
 
 struct llm_buffer {
     uint8_t * addr = NULL;

--- a/gpt4all-backend/sysinfo.h
+++ b/gpt4all-backend/sysinfo.h
@@ -2,17 +2,21 @@
 #define SYSINFO_H
 
 #include <fstream>
-#include <string>
-#include <sstream>
 #include <iomanip>
+#include <sstream>
+#include <string>
 
 #if defined(__linux__)
-#include <unistd.h>
+#   include <unistd.h>
 #elif defined(__APPLE__)
-#include <sys/types.h>
-#include <sys/sysctl.h>
+#   include <sys/types.h>
+#   include <sys/sysctl.h>
 #elif defined(_WIN32)
-#include <windows.h>
+#   define WIN32_LEAN_AND_MEAN
+#   ifndef NOMINMAX
+#       define NOMINMAX
+#   endif
+#   include <windows.h>
 #endif
 
 static long long getSystemTotalRAMInBytes()

--- a/gpt4all-backend/utils.cpp
+++ b/gpt4all-backend/utils.cpp
@@ -1,7 +1,12 @@
 #include "utils.h"
 
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
 #include <fstream>
+#include <iterator>
 #include <regex>
+#include <utility>
 
 void replace(std::string & str, const std::string & needle, const std::string & replacement) {
     size_t pos = 0;

--- a/gpt4all-backend/utils.h
+++ b/gpt4all-backend/utils.h
@@ -2,11 +2,14 @@
 
 #pragma once
 
-#include <string>
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
 #include <map>
-#include <vector>
 #include <random>
+#include <string>
 #include <thread>
+#include <vector>
 
 //
 // General purpose inline functions

